pkg/identity/identitysync/okta.go

```diff
@@ -165,6 +172,59 @@ func (o *OktaSync) ListUsers(ctx context.Context) ([]identity.IDPUser, error) {
 	return idpUsers, nil
 }
 
+func (o *OktaSync) listUsersBasedOnGroups(ctx context.Context) ([]identity.IDPUser, error) {
+	log := logger.Get(ctx)
+	idpUsers := []identity.IDPUser{}
+
+	oktaUsers := make(map[string]*okta.User)
+	userGroupIds := make(map[string][]string)
+	for _, groupId := range strings.Split(o.groups.Get(), ",") {
+		log.Infow(fmt.Sprintf("fetching users for groupId = %s", groupId))
+		users, res, err := o.client.Group.ListGroupUsers(ctx, groupId, &query.Params{})
+		if err != nil {
+			// try and log the response body
+			logResponseErr(log, res, err)
+			return nil, errors.Wrap(err, "listing okta users from okta API")
+		}
+
+		log.Debugw("listed all okta users")
+
+		for res.HasNextPage() {
+			var nextUsers []*okta.User
+			res, err = res.Next(ctx, &nextUsers)
+			if err != nil {
+				logResponseErr(log, res, err)
+				return nil, err
+			}
+			users = append(users, nextUsers...)
+			log.Debugw("fetched more users", "nextPage", res.NextPage)
+		}
+
+		for _, user := range users {
+			oktaUsers[user.Id] = user
+			userGroupIds[user.Id] = append(userGroupIds[user.Id], groupId)
+		}
+	}
+
+	// convert all Okta users to internal users
+	for _, u := range oktaUsers {
+		user, err := o.idpUserFromOktaUser(ctx, u, userGroupIds)
+		if err != nil {
+			return nil, errors.Wrapf(err, "converting okta user %s to internal user", u.Id)
+		}
```

Comment:
nit, As my other comment, put the logic of the group in the caller.

```suggestion
		user, err := o.idpUserFromOktaUser(ctx, u)
		if err != nil {
			return nil, errors.Wrapf(err, "converting okta user %s to internal user", u.Id)
		}
		user.Groups = userIdToGroupIds[u.Id]
```